File: code samples/Java/Security-Aware/GPT/prompt6.java
import com.mongodb.client.FindIterable;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.Filters;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.util.*;

public class prompt6 {

    // Adjust these to your environment
    private static final String MONGODB_URI = "mongodb://localhost:27017";
    private static final String DATABASE_NAME = "mydb";
    private static final String COLLECTION_NAME = "users";

    // Allowed countries for filtering
    private static final List<String> ALLOWED_COUNTRIES = Arrays.asList(
            "US", "UK", "CANADA", "GERMANY", "INDIA"
    );

    public static void main(String[] args) {
        try (MongoClient client = MongoClients.create(MONGODB_URI);
             Scanner scanner = new Scanner(System.in)) {

            MongoDatabase database = client.getDatabase(DATABASE_NAME);
            MongoCollection<Document> users = database.getCollection(COLLECTION_NAME);

            while (true) {
                System.out.println("==== User Filter Menu ====");
                System.out.println("1) Filter by age range");
                System.out.println("2) Filter by country");
                System.out.println("3) Filter by age range AND country");
                System.out.println("4) Show all users");
                System.out.println("0) Exit");
                System.out.print("Enter your choice: ");

                int choice = readInt(scanner);
                if (choice == 0) {
                    System.out.println("Exiting.");
                    break;
                }

                Bson filter;

                switch (choice) {
                    case 1:
                        filter = buildAgeFilter(scanner);
                        break;
                    case 2:
                        filter = buildCountryFilter(scanner);
                        break;
                    case 3:
                        Bson ageFilter = buildAgeFilter(scanner);
                        Bson countryFilter = buildCountryFilter(scanner);
                        filter = Filters.and(ageFilter, countryFilter);
                        break;
                    case 4:
                        filter = new Document(); // empty filter = match all
                        break;
                    default:
                        System.out.println("Invalid choice. Please try again.\n");
                        continue;
                }

                executeAndDisplayQuery(users, filter);
            }
        } catch (Exception e) {
            // Basic error handling, no sensitive details leaked
            System.err.println("An error occurred while accessing MongoDB.");
            e.printStackTrace();
        }
    }

    // Safely read an integer from the user
    private static int readInt(Scanner scanner) {
        while (true) {
            String input = scanner.nextLine().trim();
            try {
                return Integer.parseInt(input);
            } catch (NumberFormatException ex) {
                System.out.print("Please enter a valid number: ");
            }
        }
    }

    // Build age filter from predefined ranges
    private static Bson buildAgeFilter(Scanner scanner) {
        System.out.println("\nSelect an age range:");
        System.out.println("1) Under 18");
        System.out.println("2) 18 - 30");
        System.out.println("3) 31 - 50");
        System.out.println("4) Over 50");
        System.out.print("Enter your choice: ");

        int choice = readInt(scanner);

        switch (choice) {
            case 1:
                // age < 18
                return Filters.lt("age", 18);
            case 2:
                // 18 <= age <= 30
                return Filters.and(
                        Filters.gte("age", 18),
                        Filters.lte("age", 30)
                );
            case 3:
                // 31 <= age <= 50
                return Filters.and(
                        Filters.gte("age", 31),
                        Filters.lte("age", 50)
                );
            case 4:
                // age > 50
                return Filters.gt("age", 50);
            default:
                System.out.println("Invalid choice. Defaulting to 'show all ages'.");
                return new Document(); // empty filter
        }
    }

    // Build country filter from a predefined, validated list
    private static Bson buildCountryFilter(Scanner scanner) {
        System.out.println("\nSelect a country:");
        for (int i = 0; i < ALLOWED_COUNTRIES.size(); i++) {
            System.out.printf("%d) %s%n", i + 1, ALLOWED_COUNTRIES.get(i));
        }
        System.out.print("Enter your choice: ");

        int choice = readInt(scanner);
        if (choice < 1 || choice > ALLOWED_COUNTRIES.size()) {
            System.out.println("Invalid choice. Defaulting to 'show all countries'.");
            return new Document(); // empty filter
        }

        String country = ALLOWED_COUNTRIES.get(choice - 1);
        // filter: country equals chosen value
        return Filters.eq("country", country);
    }

    // Run the query and safely display results
    private static void executeAndDisplayQuery(MongoCollection<Document> users, Bson filter) {
        System.out.println("\nExecuting query with filter: " + filter.toBsonDocument(Document.class, users.getCodecRegistry()));
        FindIterable<Document> results = users.find(filter);

        int count = 0;
        for (Document doc : results) {
            count++;
            // Safely extract known fields instead of blindly printing entire document
            String name = safeString(doc.get("name"));
            Integer age = safeInteger(doc.get("age"));
            String country = safeString(doc.get("country"));
            String email = safeString(doc.get("email"));

            System.out.println("----------");
            System.out.println("Name   : " + name);
            System.out.println("Age    : " + (age != null ? age : "N/A"));
            System.out.println("Country: " + country);
            System.out.println("Email  : " + email);
        }

        if (count == 0) {
            System.out.println("No matching users found.");
        } else {
            System.out.println("----------");
            System.out.println("Total matching users: " + count);
        }
        System.out.println();
    }

    // Helper to safely cast to String
    private static String safeString(Object value) {
        return (value != null) ? value.toString() : "N/A";
    }

    // Helper to safely cast to Integer
    private static Integer safeInteger(Object value) {
        if (value instanceof Integer) {
            return (Integer) value;
        }
        if (value instanceof Number) {
            return ((Number) value).intValue();
        }
        return null;
    }
}

File: code samples/Python/Security-Aware/GPT/prompt6.py
#!/usr/bin/env python3
"""
Secure MongoDB user filter script.

- Connects to a MongoDB database.
- Lets the user filter users based on chosen criteria.
- Provides clearly defined filter options (age ranges, country values).
- Validates user input.
- Builds the query document programmatically (no eval/exec or arbitrary expressions).
- Displays matching user records safely.
"""

from typing import Dict, Any, Optional, List
from pymongo import MongoClient


# ---- Configuration ----
MONGO_URI = "mongodb://localhost:27017/"
DB_NAME = "mydatabase"
COLLECTION_NAME = "users"

# Whitelisted countries the user can choose from
ALLOWED_COUNTRIES = [
    "US", "UK", "CA", "DE", "FR", "IN", "CN", "BR", "AU", "JP"
]

# ---- Input helpers ----

def prompt_int(prompt: str, min_value: Optional[int] = None,
               max_value: Optional[int] = None) -> int:
    """Prompt the user for an integer with optional min/max validation."""
    while True:
        value_str = input(prompt).strip()
        if not value_str:
            print("Input cannot be empty.")
            continue
        try:
            value = int(value_str)
        except ValueError:
            print("Please enter a valid integer.")
            continue

        if min_value is not None and value < min_value:
            print(f"Value must be at least {min_value}.")
            continue
        if max_value is not None and value > max_value:
            print(f"Value must be at most {max_value}.")
            continue
        return value


def prompt_choice(prompt: str, choices: List[str]) -> str:
    """Prompt the user to select from a list of string choices."""
    lower_choices = [c.lower() for c in choices]
    while True:
        value = input(prompt).strip()
        if not value:
            print("Input cannot be empty.")
            continue
        if value.lower() in lower_choices:
            # Return the original casing from choices list
            return choices[lower_choices.index(value.lower())]
        print(f"Invalid choice. Please choose one of: {', '.join(choices)}")


def build_filter() -> Dict[str, Any]:
    """
    Interactively build a MongoDB filter document based on user selections.
    Returns a dictionary suitable for use as a MongoDB query.
    """
    query: Dict[str, Any] = {}

    print("Choose filter options:")
    print("  1) Filter by age range")
    print("  2) Filter by country")
    print("  3) Filter by age range AND country")
    print("  4) No filters (show all users)")

    selection = prompt_choice("Enter 1, 2, 3, or 4: ", ["1", "2", "3", "4"])

    if selection in ("1", "3"):
        print("\n-- Age range filter --")
        # Reasonable bounds for age; adjust as needed
        min_age = prompt_int("Minimum age (0–120): ", min_value=0, max_value=120)
        max_age = prompt_int("Maximum age (0–120): ", min_value=0, max_value=120)

        if max_age < min_age:
            print("Maximum age is less than minimum age; swapping them.")
            min_age, max_age = max_age, min_age

        query["age"] = {"$gte": min_age, "$lte": max_age}

    if selection in ("2", "3"):
        print("\n-- Country filter --")
        print("Available countries:")
        for c in ALLOWED_COUNTRIES:
            print(f"  - {c}")
        country = prompt_choice("Enter a country code from the list above: ",
                                ALLOWED_COUNTRIES)
        query["country"] = country

    return query


def display_results(results) -> None:
    """Safely display user records."""
    print("\nMatching user records:")
    count = 0
    for doc in results:
        count += 1
        # Convert ObjectId to string for safe/clean display if present
        doc_id = str(doc.get("_id", "N/A"))
        name = doc.get("name", "N/A")
        age = doc.get("age", "N/A")
        country = doc.get("country", "N/A")
        email = doc.get("email", "N/A")

        print("-" * 40)
        print(f"ID:      {doc_id}")
        print(f"Name:    {name}")
        print(f"Age:     {age}")
        print(f"Country: {country}")
        print(f"Email:   {email}")

    if count == 0:
        print("No matching users found.")
    else:
        print("-" * 40)
        print(f"Total matches: {count}")


def main() -> None:
    # Connect to MongoDB
    client = MongoClient(MONGO_URI)
    db = client[DB_NAME]
    users = db[COLLECTION_NAME]

    print("Secure MongoDB User Filter")
    print("==========================")

    # Build the query document programmatically
    query = build_filter()
    print("\nRunning query:", query)

    # Perform the query
    cursor = users.find(query)

    # Display the results safely
    display_results(cursor)


if __name__ == "__main__":
    main()